### PR TITLE
Remove Python 3.7 support

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
Per our policy, we will drop support for Python 3.7 after the next release. This PR makes appropriate changes.

## Changes proposed in this PR:
- Drop Python 3.7 support
- Deprecate / relocate `pyomo.common.backports` because it's no longer necessary
- Remove unnecessary checks / workarounds to Python 3.7

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/repn/util.py [hunk 1]
```diff
@@ -65,15 +65,14 @@ class FileDeterminism(enum.IntEnum):
     SORT_SYMBOLS = 30
 
     # We will define __str__ and __format__ so that behavior in python
-    # 3.11 is consistent with 3.7 - 3.10.
+    # 3.11 is consistent with 3.8 - 3.10.
```
Comment (jsiirola) [inline pyomo/repn/util.py:68]: This is a bit of a silly change.  The previous comment was not incorrect, and implied a longer-term behavior that we were preserving.

File: pyomo/common/backports.py [hunk 2]
```diff
@@ -1,22 +0,0 @@
...
-#  National Technology and Engineering Solutions of Sandia, LLC
-#  Under the terms of Contract DE-NA0003525 with National Technology and
-#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
-#  rights in this software.
-#  This software is distributed under the 3-clause BSD License.
-#  ___________________________________________________________________________
-
-
-# backport of contextlib.nullcontext for supporting Python < 3.7
-class nullcontext(object):
```
Comment (jsiirola) [inline pyomo/common/backports.py:14]: I would argue *against* deleting this file, as that breaks any/all potential dependencies.  Instead, this file should just
```python
from contextlib import nullcontext
```

**Review by jsiirola — CHANGES_REQUESTED** [review]
Two requested changes, and of course, the obligatory note to not merge this until after the August release.

File: pyomo/common/backports.py [hunk 3]
```diff
@@ -9,14 +9,8 @@
...
-    def __init__(self, enter_result=None):
-        self.result = enter_result
-
-    def __enter__(self):
-        return self.result
-
-    def __exit__(self, et, ev, tb):
-        return
+relocated_module_attribute(
+    'nullcontext', 'contextlib.nullcontext', version='6.6.2.dev0'
```
Comment (jsiirola) [inline pyomo/common/backports.py:15]: This should be 6.6.3.dev0 (now that the release is out)

Comment (blnicho) [reply pyomo/common/backports.py]: I think we picked 6.7.0.dev0 when we reset main for development, primarily in anticipation of this PR.

File: README.md [hunk 4]
```diff
@@ -51,7 +51,7 @@ Pyomo is available under the BSD License - see the
 
 Pyomo is currently tested with the following Python implementations:
 
-* CPython: 3.7, 3.8, 3.9, 3.10, 3.11
+* CPython: 3.8, 3.9, 3.10, 3.11
 * PyPy: 3.7, 3.8, 3.9
```
Comment (blnicho) [inline README.md:55]: Should we remove `3.7` from the PyPy line too? I don't remember how we got this list of supported PyPy versions, does it map to something in the conda-forge build?

**Review by blnicho — COMMENTED** [review]
One minor comment/question on the PyPy versions listed in the README. Also, we'll need to remove the 3.7 test from Jenkins after this gets merged.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**